File: iOS/PengChauFerry/Preferences.swift
//
//  Preferences.swift
//  PengChauFerry
//
//  Created by Lionello Lunesu on 11/4/2020.
//  Copyright © 2020 Lionello Lunesu. All rights reserved.
//

import Foundation

class Preferences {
    private static let LANGUAGE_PREF = "language"
    private static let LAST_REFRESH_PREF = "lastRefresh"

    private let shared = UserDefaults.standard

    var language: String? {
        get { shared.string(forKey: Preferences.LANGUAGE_PREF) }
        set(value) { shared.set(value, forKey: Preferences.LANGUAGE_PREF) }
    }

    var lastRefresh: LocalDateTime? {
        get { shared.string(forKey: Preferences.LAST_REFRESH_PREF).flatMap { ISO8601DateFormatter().date(from: $0) } }
        set(value) { shared.set(value.map { ISO8601DateFormatter().string(from: $0) }, forKey: Preferences.LAST_REFRESH_PREF) }
    }

}

File: Android/app/src/main/java/com/lunesu/pengchauferry/Preferences.kt
package com.lunesu.pengchauferry

import android.content.Context
import org.joda.time.LocalDateTime

class Preferences(context: Context) {
    companion object {
        private const val SHAREDPREF_NAME = "com.lunesu.pengchauferry_preferences"
        private const val LANGUAGE_PREF = "language"
        private const val LAST_REFRESH_PREF = "lastRefresh"
    }

    // Note: shared will be null during unit testing
    private val shared = context.getSharedPreferences(SHAREDPREF_NAME, Context.MODE_PRIVATE)

    var language: String?
        get() = shared.getString(LANGUAGE_PREF, null)
        set(value) = shared.edit().putString(LANGUAGE_PREF, value).apply()

    var lastRefresh: LocalDateTime?
        get() = shared.getString(LAST_REFRESH_PREF, null)?.let { LocalDateTime.parse(it) }
        set(value) { shared?.run { edit().putString(LAST_REFRESH_PREF, value?.toString()).apply() } }

}
